core/chains/terra/terratxm/txm.go

```diff
@@ -0,0 +1,355 @@
+		return
+	}
+
+	// We need to ensure that we either broadcast successfully and mark the tx as
+	// broadcasted OR we do not broadcast successfully and we do not mark it as broadcasted.
+	// We do this by first marking it broadcasted then rolling back if the broadcast api call fails.
+	var resp *txtypes.BroadcastTxResponse
+	err = txm.orm.q.Transaction(func(tx pg.Queryer) error {
+		txHash := strings.ToUpper(hex.EncodeToString(tmhash.Sum(signedTx)))
+		err = txm.orm.UpdateMsgsWithState(getSimMsgsIDs(simResults.Succeeded), Broadcasted, &txHash, pg.WithQueryer(tx))
+		if err != nil {
+			return err
+		}
+
+		txm.lggr.Infow("broadcasting tx", "from", sender, "msgs", simResults.Succeeded)
+		resp, err = txm.tc.Broadcast(signedTx, txtypes.BroadcastMode_BROADCAST_MODE_SYNC)
+		if err != nil {
+			// Rollback marking as broadcasted
+			return err
+		}
+		if resp.TxResponse == nil {
+			// Rollback marking as broadcasted
+			return errors.New("unexpected nil tx response")
+		}
+		if resp.TxResponse.TxHash != txHash {
+			// Should never happen
+			txm.lggr.Errorw("txhash mismatch", "got", resp.TxResponse.TxHash, "want", txHash)
+		}
+		return nil
+	})
+	if err != nil {
+		txm.lggr.Errorw("error broadcasting tx", "err", err, "from", sender.String())
+		// Was unable to broadcast, retry on next poll
+		return
+	}
+
+	if err := txm.confirmTx(resp.TxResponse.TxHash, getSimMsgsIDs(simResults.Succeeded)); err != nil {
+		txm.lggr.Errorw("error confirming tx", "err", err, "hash", resp.TxResponse.TxHash)
+		return
+	}
+}
+
+func (txm *Txm) confirmTx(txHash string, broadcasted []int64) error {
+	// We either mark these broadcasted txes as confirmed or errored.
+	// Confirmed: we see the txhash onchain. There are no reorgs in cosmos chains.
+	// Errored: we do not see the txhash onchain after waiting for N blocks worth
+	// of time (plus a small buffer to account for block time variance) where N
+	// is TimeoutHeight - HeightAtBroadcast. In other words, if we wait for that long
+	// and the tx is not confirmed, we know it has timed out.
+	for tries := 0; tries < txm.confirmMaxPolls; tries++ {
+		time.Sleep(txm.confirmPollPeriod)
+		// Confirm that this tx is onchain, ensuring the sequence number has incremented
+		// so we can build a new batch
+		tx, err := txm.tc.Tx(txHash)
+		if err != nil {
+			txm.lggr.Errorw("error looking for hash of tx", "err", err, "resp", txHash)
+			continue
+		}
+		// Sanity check
+		if tx.TxResponse == nil || tx.TxResponse.TxHash != txHash {
+			txm.lggr.Errorw("error looking for hash of tx, unexpected response", "tx", tx, "hash", txHash)
+			continue
+		}
+		txm.lggr.Infow("successfully sent batch", "hash", txHash, "msgs", broadcasted)
+		// If confirmed mark these as completed.
+		err = txm.orm.UpdateMsgsWithState(broadcasted, Confirmed, nil)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	// If we are unable to confirm the tx after the timeout period
+	// mark these msgs as errored
```

Comment:
Is this acceptable? That means these messages will _never_ be sent.

Reply:
yes - the intention is that we set this high enough such that it can only happen if a node accepts a tx and then fails to forward to the network. In that case we're leaning on the ocr transmission protocol to fallback to another node to try, but there is currently no gas price priority (its first come first served) so we expect that if the node accepts the tx it will get forwarded to the network.

They do plan to add gas price priority soon though, and so we're planning to shorten this block height timeout and retry with higher gas prices as follow up work for when that happens.